src/rvsdg/dag_to_graph.rs

```diff
@@ -0,0 +1,276 @@
+//! This file is a helper for translation from the dag IR to RVSDGs.
+//! It contains the `RegionGraph` struct, which is used to create a dependency graph
+//! for a region (a loop or a function body).
+//! Using the graph, we can compute a dominance frontier for if and switch statement
+//! branches.
+//! Using the dominance frontier, we decide which nodes need to be computed
+//! in the resulting region for a branch.
+
+use std::rc::Rc;
+
+use hashbrown::{HashMap, HashSet};
+use petgraph::{
+    algo::dominators::{self, Dominators},
+    graph::{DiGraph, NodeIndex},
+};
+use tree_in_context::schema::{Expr, RcExpr};
+
+pub(crate) struct RegionGraph {
+    graph: DiGraph<(), ()>,
+    expr_to_node: HashMap<*const Expr, NodeIndex>,
+    node_to_expr: HashMap<NodeIndex, Rc<Expr>>,
+    /// For each branch node, we need an extra node in the graph.
+    /// This allows us to query for the nodes dominated by the branch.
+    expr_branch_node: HashMap<(*const Expr, usize), NodeIndex>,
+    /// Dominators for the graph. This is None until the graph is fully constructed.
+    dominators: Option<Dominators<NodeIndex>>,
+}
+
+/// In the DAG IR, there are two nodes that create new "regions"
+/// by binding an argument: Function and DoWhile.
+/// region_expr should be the body of the function or the loop body.
+/// This function creates a dependency graph of all the computations for a given region
+/// (it doesn't traverse into nested regions).
+pub(crate) fn region_graph(expr: &RcExpr) -> RegionGraph {
+    let mut todo = vec![expr.clone()];
+    let mut processed = HashSet::<*const Expr>::new();
+    let mut rgraph = RegionGraph {
+        graph: DiGraph::new(),
+        expr_to_node: HashMap::new(),
+        node_to_expr: HashMap::new(),
+        expr_branch_node: HashMap::new(),
+        // dummy dominators, will be replaced later
+        dominators: None,
+    };
+    while let Some(expr) = todo.pop() {
+        if !processed.insert(Rc::as_ptr(&expr)) {
+            continue;
+        }
+        // for if or switch statements, we need to create branch nodes
+        match expr.as_ref() {
+            Expr::If(inputs, then_branch, else_branch) => {
+                let then_branch_node = rgraph.graph.add_node(());
+                rgraph
+                    .expr_branch_node
+                    .insert((Rc::as_ptr(&expr), 0), then_branch_node);
+                let else_branch_node = rgraph.graph.add_node(());
+                rgraph
+                    .expr_branch_node
+                    .insert((Rc::as_ptr(&expr), 1), else_branch_node);
+                let expr_node = rgraph.node(&expr);
+                let inputs_node = rgraph.node(inputs);
+                let then_node = rgraph.node(then_branch);
```

Comment:
I don't think I understand the difference btween `then_node` and `then_branch_node` (and same for else)?